File: code_examples/pagination.go
package pagination

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"

	"github.com/redis/go-redis/v9"
)

// PageCursor хранит сортировочные значения последней строки страницы.
type PageCursor struct {
	CreatedAt string `json:"createdAt"`
	ID        int64  `json:"id"`
}

type Post struct {
	ID        int64
	Title     string
	CreatedAt string
}

// FetchPostsPage: при nil cursor загружает первую страницу без WHERE;
// при наличии cursor делает keyset-запрос для следующей страницы.
func FetchPostsPage(ctx context.Context, db *sql.DB, cursor *PageCursor, pageSize int) ([]Post, error) {
	query := `
		SELECT id, title, created_at
		FROM posts
		ORDER BY created_at DESC, id DESC
		LIMIT $1`
	args := []any{pageSize}

	if cursor != nil {
		query = `
			SELECT id, title, created_at
			FROM posts
			WHERE (created_at, id) < ($1, $2)
			ORDER BY created_at DESC, id DESC
			LIMIT $3`
		args = []any{cursor.CreatedAt, cursor.ID, pageSize}
	}

	rows, err := db.QueryContext(ctx, query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	posts := make([]Post, 0, pageSize)
	for rows.Next() {
		var post Post
		if err := rows.Scan(&post.ID, &post.Title, &post.CreatedAt); err != nil {
			return nil, err
		}
		posts = append(posts, post)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	return posts, nil
}

// CursorFromLastRow преобразует последнюю строку результата в cursor.
func CursorFromLastRow(createdAt string, id int64) *PageCursor {
	return &PageCursor{CreatedAt: createdAt, ID: id}
}

// GetNextPostsPage: Redis хранит cursor по ключу конкретной ленты/сессии.
// Если ключа нет, это первая страница; после ответа сохраняется её граница.
func GetNextPostsPage(ctx context.Context, db *sql.DB, cache *redis.Client, feedKey string, pageSize int) ([]Post, error) {
	stored, err := cache.Get(ctx, feedKey).Result()
	var cursor *PageCursor

	if err == nil {
		cursor = &PageCursor{}
		if unmarshalErr := json.Unmarshal([]byte(stored), cursor); unmarshalErr != nil {
			return nil, fmt.Errorf("decode page cursor: %w", unmarshalErr)
		}
	} else if err != redis.Nil {
		return nil, fmt.Errorf("read page cursor: %w", err)
	}

	posts, err := FetchPostsPage(ctx, db, cursor, pageSize)
	if err != nil {
		return nil, err
	}

	if len(posts) == 0 {
		if err := cache.Del(ctx, feedKey).Err(); err != nil {
			return nil, fmt.Errorf("delete page cursor: %w", err)
		}
		return posts, nil
	}

	nextCursor := CursorFromLastRow(posts[len(posts)-1].CreatedAt, posts[len(posts)-1].ID)
	encoded, err := json.Marshal(nextCursor)
	if err != nil {
		return nil, fmt.Errorf("encode page cursor: %w", err)
	}
	if err := cache.Set(ctx, feedKey, encoded, 0).Err(); err != nil {
		return nil, fmt.Errorf("save page cursor: %w", err)
	}

	return posts, nil
}

File: code_examples/pagination.php
<?php

// Одна запись cursor: последняя строка уже загруженной страницы.
final class PageCursor
{
    public function __construct(
        public readonly string $createdAt,
        public readonly int $id,
    ) {
    }
}

// $redis - инстанс Redis, где сохраняем cursor отдельно для каждой ленты/сессии.
// Для первой страницы $cursor равен null, поэтому WHERE не добавляется.
function fetchPostsPage(PDO $db, ?PageCursor $cursor, int $pageSize = 20): array
{
    $sql = <<<'SQL'
        SELECT id, title, created_at
        FROM posts
        ORDER BY created_at DESC, id DESC
        LIMIT :page_size
    SQL;

    if ($cursor !== null) {
        $sql = <<<'SQL'
            SELECT id, title, created_at
            FROM posts
            WHERE (created_at, id) < (:cursor_created_at, :cursor_id)
            ORDER BY created_at DESC, id DESC
            LIMIT :page_size
        SQL;
    }

    $statement = $db->prepare($sql);
    $statement->bindValue(':page_size', $pageSize, PDO::PARAM_INT);

    if ($cursor !== null) {
        $statement->bindValue(':cursor_created_at', $cursor->createdAt);
        $statement->bindValue(':cursor_id', $cursor->id, PDO::PARAM_INT);
    }

    $statement->execute();
    return $statement->fetchAll(PDO::FETCH_ASSOC);
}

// Последняя строка страницы становится cursor для следующего запроса.
function cursorFromPage(array $page): ?PageCursor
{
    if ($page === []) {
        return null;
    }

    $lastPost = $page[array_key_last($page)];
    return new PageCursor($lastPost['created_at'], (int) $lastPost['id']);
}

// Главный сценарий: page 1 без cursor, следующие страницы с cursor из Redis.
function getNextPostsPage(PDO $db, Redis $redis, string $feedKey, int $pageSize = 20): array
{
    $storedCursor = $redis->get($feedKey);
    $cursor = $storedCursor === false
        ? null
        : new PageCursor(...json_decode($storedCursor, true, flags: JSON_THROW_ON_ERROR));

    $page = fetchPostsPage($db, $cursor, $pageSize);
    $nextCursor = cursorFromPage($page);

    if ($nextCursor === null) {
        $redis->del($feedKey);
    } else {
        $redis->set($feedKey, json_encode([
            'createdAt' => $nextCursor->createdAt,
            'id' => $nextCursor->id,
        ], JSON_THROW_ON_ERROR));
    }

    return $page;
}
